src/main/java/echoserver/EchoServer.java

```diff
@@ -0,0 +1,60 @@
+package echoserver;
+
+import java.io.PrintWriter;
+import java.io.IOException;
+import java.net.Socket;
+import java.util.Scanner;
+
+public class EchoServer implements Runnable {
+    private Scanner input;
+    private PrintWriter output;
+    private Socket clientSocket;
+
+    public EchoServer(Socket clientSocket) {
+        this.clientSocket = clientSocket;
+    }
+
+    public void run() {
+        try {
+            setServerInputOutput(clientSocket);
+            output.println("Connected to server");
+            echo();
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+    }
+
+    private void echo() {
+        String inputString;
+        while ((inputString = SocketIO.readFromInputStream(input)) != null) {
+            if (InputValidator.isQuit(inputString)) {
+                close();
```

Comment:
This is a good example of shutting things down through "cooperation" instead of the oft-used anti-patterns of `Thread#stop` or `System::exit`.